service/storage_proxy.cc

```diff
@@ -721,8 +742,54 @@ class storage_proxy::remote {
         }
         shared_ptr<storage_proxy> p = _sp.shared_from_this();
         auto cmd = make_lw_shared<query::read_command>(std::move(cmd1));
+
         auto src_ip = src_addr.addr;
         auto timeout = t ? *t : db::no_timeout;
+
+        auto f_r = co_await coroutine::as_future([&]() -> future<bool> {
+            // Check if we have a forward query
+            if (!cmd->slice.is_reversed())
+                return ::make_ready_future<bool>(false);
+
+            // We deal with a reversed query. Check if read command is coming from an old node.
+            // Table schema is then passed with the read command.
+            auto schema = p->local_db().find_schema(cmd->cf_id);
+            if (cmd->schema_version == schema->version()) {
+                (void)schema->get_reversed();
+                return ::make_ready_future<bool>(true);
+            }
+
+            // We deal with a reversed query. Check if read command is coming from an new node.
+            // Reversed table schema is then passed with the read command.
+            if (cmd->schema_version == reversed(schema->version()))
+                return ::make_ready_future<bool>(false);
+
+            // The table schema is outdated. Still do not know if the read command comes from
+            // and old or a new new. Ask coordinator for a schema.
+            return get_schema_for_read(cmd->schema_version, std::move(src_addr), timeout).then_wrapped([&](auto f){
+                if (f.failed()) {
+                    return get_schema_for_read(reversed(cmd->schema_version), std::move(src_addr), timeout);
+                }
+
+                return ::make_ready_future<schema_ptr>(f.get());
+            }).then([&cmd](auto schema){
+                // A schema is received from the coordinator. Check if it is a table schema.
+                if (schema->version() != cmd->schema_version) {
```

Comment:
I don't think this is enough to determine whether the coordinator is old or new, we need the first two checks from this method. Maybe move those into a separate function and invoke them from here as well.

Reply:
You are right! Good catch. My comment above is completely misleading.